x-pack/plugins/elastic_assistant/server/lib/attack_discovery/graphs/default_attack_discovery_graph/nodes/generate/index.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ActionsClientLlm, AssistantToolLlm } from '@kbn/langchain/server';
+import type { Logger } from '@kbn/core/server';
+
+import { discardPreviousGenerations } from './helpers/discard_previous_generations';
+import { extractJson } from '../helpers/extract_json';
+import { getAnonymizedAlertsFromState } from './helpers/get_anonymized_alerts_from_state';
+import { getChainWithFormatInstructions } from '../helpers/get_chain_with_format_instructions';
+import { getCombined } from '../helpers/get_combined';
+import { getCombinedAttackDiscoveryPrompt } from '../helpers/get_combined_attack_discovery_prompt';
+import { generationsAreRepeating } from '../helpers/generations_are_repeating';
+import { getUseUnrefinedResults } from './helpers/get_use_unrefined_results';
+import { parseCombinedOrThrow } from '../helpers/parse_combined_or_throw';
+import { responseIsHallucinated } from '../helpers/response_is_hallucinated';
+import type { GraphState } from '../../types';
+
+export const getGenerateNode = ({
+  llm,
+  logger,
+}: {
+  llm: ActionsClientLlm | AssistantToolLlm;
+  logger?: Logger;
+}): ((state: GraphState) => Promise<GraphState>) => {
+  const generate = async (state: GraphState): Promise<GraphState> => {
+    logger?.debug(() => `---GENERATE---`);
```

Comment:
Nit to remove of move into the log on L57. The rest make sense since they are conditional

Reply:
When I tried to move this, there didn't appear to be an elegant way to combine L31 and L37 because the intent of `debug` logging at the entry point of generate is to confirm that the node was called, even if it throws. That said, `llmType` is required when logging on L57 to disambiguate when multiple graphs are running simultaneously during evaluations, but driving `llmType` can throw.